server.js

```diff
@@ -7395,6 +7395,148 @@ cache(function(data, match, sendBadge, request) {
   });
 }));
 
+// PHP version from .travis.yml
+camp.route(/^\/travis-yml\/php-v\/([^/]+\/[^/]+)(?:\/([^/]+))?\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var userRepo = match[1];  // eg, espadrine/sc
+  var version = match[2];
+  var format = match[3];
+  if (!version) {
+    version = 'master';
+  }
+  var php_releases = {
+    '5': [2, 3, 4, 5, 6],
+    '7': [0, 1, 2],
+  };
+  var options = {
+    method: 'GET',
+    uri: 'https://api.travis-ci.org/repos/' + userRepo + '/branches/' + version,
+  };
+  var badgeData = getBadgeData('PHP', data);
+  request(options, function(err, res, buffer) {
+    if (err != null) {
+      log.error('Travis CI error: ' + err.stack);
+      if (res) {
+        log.error('' + res);
+      }
+      badgeData.text[1] = 'invalid';
+      sendBadge(format, badgeData);
+      return;
+    }
+
+    try {
+      var data = JSON.parse(buffer);
+      var has_hhvm = false;
+      var versions = {};
+      var isEqual = function (array1, array2) {
+        return array1.length === array2.length && array1.every(function(v, i) {
+          return v === array2[i];
+        });
+      };
+      var parseVersion = function (version) {
+        if (version === 'nightly') {
+          // ignore nightly builds
+        } else if (version.indexOf('hhvm') === 0) {
+          has_hhvm = true;
+        } else {
+          var numbers = version.split('.');
+          if (!(numbers[0] in versions)) {
+            versions[numbers[0]] = [];
+          }
+          if (typeof numbers[1] !== 'undefined') {
+            numbers[1] = parseInt(numbers[1]);
+          } else {
+            numbers[1] = 0;
+          }
+          if (versions[numbers[0]].indexOf(numbers[1]) === -1) {
+            versions[numbers[0]].push(numbers[1]);
+            versions[numbers[0]].sort();
+          }
+        }
+      };
+
+      // build versions map
+      if (typeof data.branch.config.php !== 'undefined') {
+        for (var php_index in data.branch.config.php) {
+          parseVersion(data.branch.config.php[php_index].toString());
+        }
+      }
+      if (typeof data.branch.config.matrix.include !== 'undefined') {
+        for (var matrix_index in data.branch.config.matrix.include) {
+          parseVersion(data.branch.config.matrix.include[matrix_index].php.toString());
+        }
+      }
```

Comment:
This is difficult to follow.

There are some functions in [lib/php-version.js](https://github.com/badges/shields/blob/master/lib/php-version.js) which you could take advantage of. Could you extract the bulk of this code to additional functions in that file, and add tests which convey + document what they do?

There are some good examples of tests in [lib/luarocks-version.spec.js](https://github.com/badges/shields/blob/master/lib/luarocks-version.spec.js).

Reply:
I've already thought about this. will do